feat(gatsby-remark-images): make images blur up

jamesadarich opened the pull request:
Hey guys,

This is a proposal for adding functionality to mimic the blur up effect in `gatsby-image` for `gatsby-remark-images`. This could be nicer I think but I can't see an obvious way to do that without a great deal of refactoring around how the remark plugins are integrating (but hopefully you guys can guide to an awesome solution :) )

The pros and cons of this solution, I believe, are as follow;

**Pros**
* in keeping with current code (should be minimal impact change)
* works for images with transparencies
* no need to calculate ratio

**Cons**
* inline script - not great for CSP but I've had some thoughts about resolutions for this in general (but this is a separate piece)
* reliant on class names staying the same (but I did a minor refactor to help avoid this being a problem)
* not really testable (I have tested against a local gatsby solution and it seems to work cross browser :) )

Let me know what you guys think about the solution and any ideas you guys have

KyleAMathews commented:
Looks promising!

Instead of inlining the code however (which would cause a lot of code duplication), you can add the code using gatsby-ssr.js to the `<head>` using `onRenderBody` https://next.gatsbyjs.org/docs/ssr-apis/#onRenderBody

jamesadarich commented:
@KyleAMathews I've given this a go but under tests it would seem that `gatsby` hasn't loaded the markdown at the point `onRenderBody` adds the script so this is proving problematic. From reading the other available API docs doesn't seem to be anything that would cover this situation unless you've got any ideas? :)

KyleAMathews commented:
This is for server rendering. Check out some of the other components that use this?

Another option is to use https://www.gatsbyjs.org/docs/browser-apis/#onRouteUpdate

jamesadarich commented:
Newest update moves script to `gatsby-browser.js` as expected. Behaviour still remains the same except for internet explorer which seems to fail as the javascript isn't compiled to es5 in the bundle. Need to investigate this one a bit more though.

jamesadarich commented:
Spot on 👍 all up to date now ready for another review round :)

jamesadarich commented:
@KyleAMathews I've removed the [WIP] tag as I think we've gotten to a good point now.

Final improvements:
* if image loads form cache the effect does **not** show as we already have the image ready for the user
* if for whatever reason the script fails there is some defensive code in there to only set the opacity of unloaded images to 0 after we know we can implement the effect

I've updated the snapshot in line with the changes so we should be ready to go barring any feedback you guys have :) let me know if you want any tweaks.

As a side note I found when using a for of loop, babel didn't seem to be transpiling this to valid es5 (it uses Symbol which is es2015) is this an issue that needs to be picked up separately?

pieh commented on packages/gatsby-remark-images/src/index.js:
```diff
@@ -103,12 +100,17 @@ module.exports = (
+      left: 0;
+      transition: opacity 0.5s;
+      transition-delay: 0.5s;
+      box-shadow: inset 0px 0px 0px 400px ${
```
Was making this verbose intentional? This inflates outputted html with extra bytes. Perhaps we can keep it in readable form but actually output compressed code?

jamesadarich replied:
Sure, my assumption was that this was minified down so shouldn't make a difference but much easier to understand what's going on.

pieh commented on packages/gatsby-remark-images/src/index.js:
```diff
@@ -103,12 +100,17 @@ module.exports = (
+      top: 0;
+      left: 0;
+      transition: opacity 0.5s;
+      transition-delay: 0.5s;
```
we can merge those into single `transition: opacity 0.5s 0.5s;`

pieh replied:
maybe we could move this to that that `classes.js` file (maybe rename it to `constants.js`) where we could have it defined verbose there, but export actually compressed version (with removed all white spaces)? and only dynamically add `box-shadow` property here (as it depends on `backgroundColor`)?

something like:
```js
const imageStyle = `
      width: 100%;
      height: 100%;
      margin: 0;
      vertical-align: middle;
      position: absolute;
      top: 0;
      left: 0;
      transition: opacity 0.5s;
      transition-delay: 0.5s;
`

const compress = css => {
    const ruleExtractor = /(\S+)\s*:\s*(.+);/g

    let rules = ''
    while ((matches = ruleExtractor.exec(imageStyle)) !== null) {
        rules += `${matches[1]}:${matches[2]};`
    }
    return rules
}

exports.compressedImageStyle = compress(css)
```
which should export `width:100%;height:100%;margin:0;vertical-align:middle;position:absolute;top:0;left:0;transition:opacity 0.5s;transition-delay:0.5s;`

pieh replied:
Also just realised - there shouldn't be delay on main image transition I think - it's only useful on placeholder

oorestisime commented:
Just tested this locally and indeed images are loaded as soon as we navigate into the page. I am wondering if there is anyway to use gatsby-image directly here since it will simplify the intersectionObserver part (and maybe styling etc)

jamesadarich commented:
@pieh to clarify on this one, are we expecting to request the images when the page loads and then only show when the observer fires. Or only request the image when the observer fires?

jamesadarich commented:
@oorestisime had a look into using `gatsby-image` directly but as the remark requires a string to be passed to it we can't use it directly, but certainly gain inspiration :)

jamesadarich commented:
Hey @pieh, I've done a bit of looking at `gatsby-image` and it does not request the image until the intersection observer is triggered. To achieve this we'd have to lift a lot of the heavy work to the browser which may be a bit painful, until we can have remark handling react and use `gatsby-image` I'd suggest we just use the observer to perform the effect on scroll because we won't have degraded performance on the browser (images are loaded immediately as they are) and by trying to introduce this we're adding more browser complexity, which brings behaviour regression risk but also increase in load for the browser and finally it will bring the look and feel of `gatsby-remark-image` inline with `gatsby-image`

If you're in agreement I'll implement the intersection observer to maintain the request behavior but add the blur up on visible behavior

pieh commented:
I wouldn't suggest using `gatsby-image` here. For this to work users would need to use https://github.com/rhysd/rehype-react (definitely possible, but this is pretty niche + definitely more demanding than rendering html strings and adding some js).

I would just like same approach as in `gatsby-image` of using IntersectionObserver, so we load full res images that user can actually see (to limit how many request we are making, so images users see actually load faster).

jamesadarich commented:
Cool, sounds like a great idea @pieh, if I add the review feedback plus basic intersection observer functionality so we can get this functionality cleared up and available. I'm happy to raise a separate issue to work on the full intersection functionality I feel like it'd be great to get this piece available soon as it's been sitting for a while and it is a good improvement. Does that work for you?

pieh commented:
@jamesadarich sorry, I've missed question. I'm not sure what You mean with "full intersection functionality"? 

Only reason to have IntersectionObserver is to start loading image when user would actually see it. There is not much benefit to have it just for fade animation - it's better to show full resolution image as soon as it's available.

jamesadarich commented:
@pieh that's cool if we think there's no benefit to implement it purely for visuals. The goal of this PR was to keep the load functionality the same but improve the end user experience to be more in line with `gatsby-image` as noted in #7781.

I agree that the IntersectionObserver would be awesome to get in but this would be a improvement to loading more than implementing the animation. But based on all the different DOM that can be output from this plugin I wanted to minimize the impact and get some benefit that users are looking for so that the IntersectionObserver can be implemented at a later date.

The main points this PR solves are;

1. enabling use of images with transparency with the "blur up" functionality (currently if you do this when the image loads you can see the placeholder underneath it)
2. have the blur up effect actually implemented (currently the image is just loaded over the top of the placeholder which can result in the "creeping" style loading)
3. bring it closer to `gatsby-image` in behaviour

If we're happy that this brings enough benefit on it's own I'll implement the remaining requests for change that you suggested :)

pieh commented on packages/gatsby-remark-images/src/gatsby-browser.js:
```diff
@@ -0,0 +1,26 @@
+        // set the opacity to zero after we identify matching images
+        // to avoid issues where this script loads incorrectly
+        // resulting in hidden images
+        imageElement.style.opacity = 0
```
I don't understand why you need to set opacity to 0 here

jamesadarich replied:
the reason is to avoid breaking things when unexpected behaviour happens e.g. disabled JavaScript to make sure that the image is still visible in this case

pieh replied:
Won't this cause to be some weirdness with opacity transition?

pieh replied:
what about removing `transition` from html markup and adding it here? Also we should only play with `opacity` if image is not `complete` I think? Something like this:

```js
const onImageLoad = () => {
  imageElement.style.opacity = 0
  imageElement.style.transition = `opacity 0.5s`
  backgroundElement.style.opacity = 0
  imageElement.removeEventListener(`load`, onImageLoad)
}

if (image.complete) {
  backgroundElement.style.opacity = 0
} else {
  imageElement.style.opacity = 0
  imageElement.addEventListener(`load`, onImageLoad)
}
```

jamesadarich replied:
Not that I noticed when I tested it but happy to remove if you're concerned

pieh replied:
I did test it and with `transition` typo below there was definitely some glitches :)

jamesadarich replied:
sorry commented before your second comment. I'll have a tinker based on your suggestion :)

pieh commented on packages/gatsby-remark-images/src/gatsby-browser.js:
```diff
@@ -0,0 +1,26 @@
+
+        const imageLoadHandler = createImageLoadHandler(backgroundElement, imageElement)
+
+        imageElement.complete ? imageLoadHandler() : imageElement.addEventListener(`load`, imageLoadHandler)
```
you never remove EventListener

jamesadarich replied:
great catch 👍 I'll fix :)

pieh commented on packages/gatsby-remark-images/src/index.js:
```diff
@@ -171,17 +172,17 @@ module.exports = (
-    >${imageTag}</span>
+    <img
+      class="${imageBackgroundClass}"
+      style="width: 100%; height: 100%; position: relative; bottom: 0; left: 0; transition-delay: 0.5s; tranistion: opacity 0.5s; opacity: 1; background-size: cover; display: block;"
```
typo in `transition` property

pieh replied:
I'm also not sure if it's good idea changing span to img here - there might be a11y implications if screenreader see 2 image tags instead of one for essentially same image?

jamesadarich replied:
Hmmm... not sure to be honest I'll take your guidance on this one. Looking at `gatsby-image` it seems that the actual image is wrapped in a Tag that's passed in as a prop. So would we put this back into the span and keep the background as is? As this would be similar to what's going on there.

jamesadarich replied:
From memory it was while I was working through a good way to implement it so went through several variations so I think this is probably just a side effect. I'll add it back in so essentially the below?

```
<img class="${imageBackgroundClass}" ... />
<span>${imageTag}</span>
```

pieh replied:
so previously it was
```html
<span class="gatsby-resp-image-wrapper">
  <span class="gatsby-resp-image-background-image">
    <img>
  </span>
</span>
```
which won't work if we won't to hide placeholder that previously was in `.gatsby-resp-image-background-image` (as it would hide img too) maybe this:
```html
<span class="gatsby-resp-image-wrapper">
  <span class="gatsby-resp-image-background-image"></span>
  <img>
</span>
```
to change as little as possible (limit potential regressions) in this PR?

jamesadarich replied:
I'll have a play around to see what works 👍 thanks for comments, appreciate it :)

pieh commented on packages/gatsby-remark-images/src/gatsby-browser.js:
```diff
@@ -0,0 +1,26 @@
+exports.onRouteUpdate = () => {    
+    const imageWrappers = document.querySelectorAll(`.${imageWrapperClass}`)
+
+    Array.prototype.forEach.call(imageWrappers, (imageWrapper) => {
```
I think it would be much more readable if you would use:

```
const imageWrappers = Array.from(document.querySelectorAll(`.${imageWrapperClass}`))

imageWrappers.forEach(imageWrapper => {
```

jamesadarich replied:
I remember why I didn't do this. There is no polyfill currently loaded and felt probably wasn't worthwhile in this case since readability was similar. Do you think it's worth adding a polyfill or keep as is @pieh?

KyleAMathews commented:
Yeah, IntersectionObserver is great as it's a cheap way to implement lazy loading — so definitely great but can come in a follow-up PR

jamesadarich commented:
@pieh 
1.
yep, true I missed that one sorry. Just so long since I originally did that I forgot I'll sort that :)

2.
a) any guidance on where this should be done?
b) is this going to block the PR or should be something raised as an issue to resolve separately?

pieh commented:
> a) any guidance on where this should be done?
> b) is this going to block the PR or should be something raised as an issue to resolve separately?

I mean this could be just code comment right above you using it - it would be great if this would something like article we could link to (i.e. https://css-tricks.com/snippets/javascript/loop-queryselectorall-matches/ ). Which btw it seems like using classic `for` loop might be best comprise between readable and supported without polyfill

jamesadarich commented:
true, I'll refactor to `for of`

pieh commented:
published `gatsby-remark-images@3.0.0`

t2ca commented:
Would be great if someone could make gatsby-remark-images-contentful compatible with this update.

Thanks!

KyleAMathews commented:
This is awesome! Great work @jamesadarich!

jamesadarich commented:
@KyleAMathews thanks, loving Gatsby glad I could give something back :)

@pieh thanks for your patience and assistance with this one 👍 

@t2ca I've just had a brief look at this package and has a lot of stuff in common so probably worth sharing some of the functionality for sure. Maybe we could talk about how to do that if it's something you're keen on doing? Feel free to contact me directly :) Like with @oorestisime it's not something I use but it'd be cool to refactor the code to have more shared stuff so it's interesting but I think lazy loading for me is higher priority

@oorestisime I definitely did want to have a peek at this but maybe it's something we can collaborate on if that works for you? :)

oorestisime commented:
nah go ahead! my intention was just to make sure this gets done. it is not urgent, for me, so go for it. (i can have a look/test it once you have something :) )
